File: src/moe_nexus/cache_engine/tokenizer.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Optional, Union

import numpy as np
import torch


@dataclass
class NumberTokenizer:
    vocab_size: int = 256
    pad_token: str = "<pad>"
    unk_token: str = "<unk>"
    bos_token: str = "<bos>"
    eos_token: str = "<eos>"

    def __post_init__(self) -> None:
        self._char_to_int: Dict[str, int] = {}
        self._int_to_char: Dict[int, str] = {}
        self._setup_default_vocab()

    def _setup_default_vocab(self) -> None:
        special_tokens = [self.pad_token, self.unk_token, self.bos_token, self.eos_token]
        for i, token in enumerate(special_tokens):
            self._char_to_int[token] = i
            self._int_to_char[i] = token
        for i in range(256):
            char = chr(i)
            if char not in self._char_to_int:
                idx = len(self._char_to_int)
                self._char_to_int[char] = idx
                self._int_to_char[idx] = char

        self.vocab_size = len(self._char_to_int)
        self.pad_token_id = self._char_to_int[self.pad_token]
        self.unk_token_id = self._char_to_int[self.unk_token]
        self.bos_token_id = self._char_to_int[self.bos_token]
        self.eos_token_id = self._char_to_int[self.eos_token]

        self._lookup_array = np.zeros(max(self._int_to_char.keys()) + 1, dtype=np.uint32)
        for idx, char in self._int_to_char.items():
            if len(char) == 1:
                self._lookup_array[idx] = ord(char)
            else:
                self._lookup_array[idx] = 0

        # Wektoryzowana tablica kodowania: bajt -> token id.
        self._encode_lut = np.array(
            [self._char_to_int.get(chr(b), self.unk_token_id) for b in range(256)],
            dtype=np.int32,
        )

    def encode(self, text: str, add_bos: bool = False, add_eos: bool = False) -> List[int]:
        tokens: List[int] = []
        if add_bos:
            tokens.append(self.bos_token_id)
        if text.isascii():
            arr = np.frombuffer(text.encode("ascii"), dtype=np.uint8)
            tokens.extend(self._encode_lut[arr].tolist())
        else:
            for char in text:
                tokens.append(self._char_to_int.get(char, self.unk_token_id))
        if add_eos:
            tokens.append(self.eos_token_id)
        return tokens

    def encode_tensor(self, text: str, add_bos: bool = False, add_eos: bool = False) -> torch.Tensor:
        tokens = self.encode(text, add_bos=add_bos, add_eos=add_eos)
        return torch.tensor(tokens, dtype=torch.long)

    def decode(self, tokens: Union[List[int], torch.Tensor, np.ndarray]) -> str:
        if isinstance(tokens, torch.Tensor):
            tokens = tokens.cpu().numpy().tolist()
        elif isinstance(tokens, np.ndarray):
            tokens = tokens.tolist()
        chars = []
        for token in tokens:
            if token in self._int_to_char:
                char = self._int_to_char[token]
                if char in (self.pad_token, self.unk_token, self.bos_token, self.eos_token):
                    continue
                chars.append(char)
        return "".join(chars)

    def batch_encode(self, texts: List[str], max_length: Optional[int] = None, pad: bool = True) -> torch.Tensor:
        encoded = [self.encode_tensor(text, add_bos=False, add_eos=False) for text in texts]
        if pad and max_length is None:
            max_length = max(t.size(0) for t in encoded)
        elif not pad:
            return torch.nn.utils.rnn.pad_sequence(encoded, batch_first=True)
        padded = []
        for t in encoded:
            if t.size(0) < max_length:
                pad_size = max_length - t.size(0)
                t = torch.cat([t, torch.full((pad_size,), self.pad_token_id, dtype=torch.long)])
            padded.append(t[:max_length])
        return torch.stack(padded)

    def get_vocab_size(self) -> int:
        return self.vocab_size

    def get_lookup_array(self) -> np.ndarray:
        return self._lookup_array

File: cpp/src/tokenizer.cpp
#include "moe_nexus/tokenizer.h"
#include <stdexcept>

namespace moe_nexus {

NumberTokenizer::NumberTokenizer(const TokenizerConfig& config) : config_(config) {
    build_vocab();
    build_lookup_array();
}

void NumberTokenizer::build_vocab() {
    char_to_int_.clear();
    int_to_char_.clear();
    
    // Special tokens first
    std::vector<std::pair<std::string, int>> special_tokens = {
        {config_.pad_token, 0},
        {config_.unk_token, 1},
        {config_.bos_token, 2},
        {config_.eos_token, 3}
    };
    
    for (const auto& [token, id] : special_tokens) {
        char_to_int_[token] = id;
        int_to_char_[id] = token;
    }
    
    // ASCII characters
    for (int i = 0; i < 256; ++i) {
        char c = static_cast<char>(i);
        std::string s(1, c);
        if (char_to_int_.find(s) == char_to_int_.end()) {
            int id = static_cast<int>(char_to_int_.size());
            char_to_int_[s] = id;
            int_to_char_[id] = s;
        }
    }
    
    vocab_size_ = static_cast<int>(char_to_int_.size());
    pad_token_id_ = char_to_int_.at(config_.pad_token);
    unk_token_id_ = char_to_int_.at(config_.unk_token);
    bos_token_id_ = char_to_int_.at(config_.bos_token);
    eos_token_id_ = char_to_int_.at(config_.eos_token);
}

void NumberTokenizer::build_lookup_array() {
    size_t max_id = 0;
    for (const auto& [id, str] : int_to_char_) {
        if (static_cast<size_t>(id) > max_id) {
            max_id = static_cast<size_t>(id);
        }
    }
    
    lookup_array_.resize(max_id + 1, 0);
    for (const auto& [id, str] : int_to_char_) {
        if (str.size() == 1) {
            lookup_array_[static_cast<size_t>(id)] = static_cast<uint32_t>(str[0]);
        }
    }
}

std::vector<int> NumberTokenizer::encode(const std::string& text, bool add_bos, bool add_eos) const {
    std::vector<int> tokens;
    tokens.reserve(text.size() + 2);
    
    if (add_bos) {
        tokens.push_back(bos_token_id_);
    }
    
    for (char c : text) {
        std::string s(1, c);
        auto it = char_to_int_.find(s);
        if (it != char_to_int_.end()) {
            tokens.push_back(it->second);
        } else {
            tokens.push_back(unk_token_id_);
        }
    }
    
    if (add_eos) {
        tokens.push_back(eos_token_id_);
    }
    
    return tokens;
}

std::string NumberTokenizer::decode(const std::vector<int>& tokens) const {
    return decode(tokens.data(), tokens.size());
}

std::string NumberTokenizer::decode(const int* tokens, size_t length) const {
    std::string result;
    result.reserve(length);
    
    for (size_t i = 0; i < length; ++i) {
        int id = tokens[i];
        if (id < 0 || static_cast<size_t>(id) >= lookup_array_.size()) {
            continue;
        }
        
        uint32_t cp = lookup_array_[static_cast<size_t>(id)];
        if (cp == 0) continue; // Skip special tokens
        
        result.push_back(static_cast<char>(cp));
    }
    
    return result;
}

} // namespace moe_nexus
